rust/src/jsonbuilder.rs

```diff
@@ -651,38 +699,45 @@ impl JsonBuilder {
                     &buf[0..offset],
                     val.as_bytes(),
                 );
+                self.buf.try_reserve(error.len())?;
                 self.buf.push_str(&error);
             }
         }
         Ok(())
     }
 }
 
-/// A Suricata specific function to create a string from bytes when UTF-8 decoding fails.
+/// Create a string from bytes, using \xXX respresentation for bytes
+/// over 0x0f.
 ///
-/// For bytes over 0x0f, we encode as hex like "\xf2".
-fn string_from_bytes(input: &[u8]) -> String {
-    let mut out = String::with_capacity(input.len());
+/// A TryReserveError is returned if the string cannot be allocated.
+fn try_string_from_bytes(input: &[u8]) -> Result<String, TryReserveError> {
+    let mut out = String::new();
+    out.try_reserve(input.len() * 4)?;
     for b in input.iter() {
         if *b < 128 {
             out.push(*b as char);
         } else {
             out.push_str(&format!("\\x{:02x}", *b));
         }
     }
-    return out;
+    return Ok(out);
 }
 
 #[no_mangle]
 pub extern "C" fn jb_new_object() -> *mut JsonBuilder {
-    let boxed = Box::new(JsonBuilder::new_object());
-    Box::into_raw(boxed)
+    match JsonBuilder::try_new_object() {
+        Ok(js) => Box::into_raw(Box::new(js)),
```

Comment:
is `Box::new()` also doing an alloc?

Reply:
It is.  `Box::try_new` is an experimental API that would be ideal here, but until then it should probably be wrapped in a panic handler.  I need to actually look into if Rust memory allocations panic or abort, the latter not being catchable.